### PR TITLE
case 20701: fixes for grab

### PR DESCRIPTION
**Pull request by AndrewMeadows**
https://highfidelity.manuscript.com/f/cases/20701/Simulation-ownership-is-not-transferred-when-grabbing-resulting-in-frequently-jittery-grabbed-entities

https://highfidelity.manuscript.com/f/cases/20815/non-dynamic-entities-don-t-collide-correct-when-grabbed

This PR for a few grab fixes.

### CODE REVIEW

File: libraries/avatars-renderer/src/avatars-renderer/Avatar.cpp [hunk 1]
```diff
@@ -366,25 +366,8 @@ bool Avatar::updateGrabs() {
...
                 target->removeGrab(grab);
                 delItr.remove();
-                // in case this is the last grab on an entity, we need to shrink the queryAACube and tell the server
-                // about the final position.
-                if (entityTree) {
-                    bool force = true;
-                    entityTree->withWriteLock([&] {
-                        entityTree->updateEntityQueryAACube(target, packetSender, force, iShouldTellServer);
-                    });
-                }
```
Comment (AndrewMeadows) [inline libraries/avatars-renderer/src/avatars-renderer/Avatar.cpp]: This is unnecessary.  It is the simulation owner's responsibility will supply the correct queryAACube.  The entity-sever requires constant feedings for this property because it doesn't have enough information to compute it.  No, really, it's true!

Meanwhile there is magic code on the receiving side to ignore transform updates from the servers when grabs are involved.  That is a different story.

File: libraries/avatars-renderer/src/avatars-renderer/Avatar.cpp [hunk 2]
```diff
@@ -395,12 +378,22 @@ bool Avatar::updateGrabs() {
...
             SpatiallyNestablePointer target = SpatiallyNestable::findByID(grab->getTargetID(), success);
 
             if (success && target) {
                 target->addGrab(grab);
+                if (isMyAvatar()) {
+                    EntityItemPointer entity = std::dynamic_pointer_cast<EntityItem>(target);
+                    if (entity) {
+                        entity->upgradeScriptSimulationPriority(PERSONAL_SIMULATION_PRIORITY);
+                    }
+                }
```
Comment (AndrewMeadows) [inline libraries/avatars-renderer/src/avatars-renderer/Avatar.cpp]: `MyAvatar` needs to bid for simulation ownership whenever it adds a grab.

File: libraries/avatars-renderer/src/avatars-renderer/Avatar.cpp [hunk 3]
```diff
@@ -395,12 +378,22 @@ bool Avatar::updateGrabs() {
         while (changeItr.hasNext()) {
             QUuid grabID = changeItr.next();
             GrabPointer& grab = _avatarGrabs[grabID];
+            if (!grab) {
+                changeItr.remove();
+                continue;
+            }
```
Comment (AndrewMeadows) [inline libraries/avatars-renderer/src/avatars-renderer/Avatar.cpp]: This fixes a potential resource leak: otherwise there was no way for an entry in `_changedAvatarGrabs` to be removed if it could not be found in `_avatarGrabs`.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16713/HighFidelity-Beta-PR14766-aff1244.apk

File: libraries/entities/src/EntityItem.cpp [hunk 4]
```diff
@@ -3360,7 +3360,7 @@ void EntityItem::addGrab(GrabPointer grab) {
     enableNoBootstrap();
     SpatiallyNestable::addGrab(grab);
 
-    if (getDynamic() && getParentID().isNull()) {
+    if (_physicsInfo && getParentID().isNull()) {
```
Comment (AndrewMeadows) [inline libraries/entities/src/EntityItem.cpp]: This change allows us to fall into the "use real Action" path for any grabbed Entity that has a Body in the physics simulation.  Non-dynamic things will get an `AvatarActionHold` instance that will measure effective velocities and apply them to the Body (if it can: non-zero velocities don't actually work on concave Bullet shapes, but I believe they fail "gently").

File: libraries/entities/src/EntityItem.cpp [hunk 5]
```diff
@@ -3398,11 +3398,14 @@ void EntityItem::addGrab(GrabPointer grab) {
...
         _grabActions[actionID] = action;
         simulation->addDynamic(action);
+        markDirtyFlags(Simulation::DIRTY_MOTION_TYPE);
+        simulation->changeEntity(getThisPointer());
     }
 }
 
 void EntityItem::removeGrab(GrabPointer grab) {
     SpatiallyNestable::removeGrab(grab);
+    markDirtyFlags(Simulation::DIRTY_MOTION_TYPE);
```
Comment (AndrewMeadows) [inline libraries/entities/src/EntityItem.cpp]: Set the `DIRTY_MOTION_TYPE` bit whenever we add/remove a grab and also put it on the "chaned entity" list (for the `addGrab()` case; the `removeGrab()` case already worked through some mysterious path).  This lets the physics simulation know the object might need to transition from `STATIC` to `KINEMATIC` or visa versa.

File: libraries/entities/src/SimulationOwner.h [hunk 6]
```diff
@@ -96,11 +96,11 @@ const uint8_t RECRUIT_SIMULATION_PRIORITY = VOLUNTEER_SIMULATION_PRIORITY + 1;
 // When poking objects with scripts an observer will bid at SCRIPT_EDIT priority.
 const uint8_t SCRIPT_GRAB_SIMULATION_PRIORITY = 128;
 const uint8_t SCRIPT_POKE_SIMULATION_PRIORITY = SCRIPT_GRAB_SIMULATION_PRIORITY - 1;
-const uint8_t AVATAR_ENTITY_SIMULATION_PRIORITY = 255;
 
 // PERSONAL priority (needs a better name) is the level at which a simulation observer owns its own avatar
 // which really just means: things that collide with it will be bid at a priority level one lower
 const uint8_t PERSONAL_SIMULATION_PRIORITY = SCRIPT_GRAB_SIMULATION_PRIORITY;
+const uint8_t AVATAR_ENTITY_SIMULATION_PRIORITY = PERSONAL_SIMULATION_PRIORITY;
```
Comment (AndrewMeadows) [inline libraries/entities/src/SimulationOwner.h]: I realized after auditing this code that setting the `AVATAR_ENTITY_SIMULATION_PRIORITY` so high it would cause some bad effects with the "ownership infection" logic.  This is the simple easy fix.  Maybe we'll revisit this someday.  I think the problem here is the _meaning_ of the the `simulation ownership priority` value is interpreted in multiple ways, and values above 128 just don't make sense right now for some of those ways.

In other words: this change fixes a bug no one was aware of yet: use an avatar entity as a bat and you could gain simulation ownership of any other non-avatar-entity no matter how someone else was using it.

File: libraries/physics/src/EntityMotionState.cpp [hunk 7]
```diff
@@ -211,6 +211,7 @@ PhysicsMotionType EntityMotionState::computePhysicsMotionType() const {
     }
     if (_entity->isMovingRelativeToParent() ||
         _entity->hasActions() ||
+        _entity->hasGrabs() ||
         _entity->hasAncestorOfType(NestableType::Avatar)) {
         return MOTION_TYPE_KINEMATIC;
     }
```
Comment (AndrewMeadows) [inline libraries/physics/src/EntityMotionState.cpp]: `EntityMotionState::computePhysicsMotionType()` needs to report `MOTION_TYPE_KINEMATIC` for non-dynamic Entities that are grabbed.

File: libraries/physics/src/EntityMotionState.cpp [hunk 8]
```diff
@@ -767,6 +768,7 @@ bool EntityMotionState::shouldSendBid() const {
     // NOTE: this method is only ever called when the entity's simulation is NOT locally owned
     return _body->isActive()
         && (_region == workload::Region::R1)
+        && _ownershipState != EntityMotionState::OwnershipState::Unownable
         && glm::max(glm::max(VOLUNTEER_SIMULATION_PRIORITY, _bumpedPriority), _entity->getScriptSimulationPriority()) >= _entity->getSimulationPriority()
         && !_entity->getLocked();
 }
```
Comment (AndrewMeadows) [inline libraries/physics/src/EntityMotionState.cpp]: This is an optimization and a safety check: avoid bidding for things you can never own.

File: libraries/physics/src/EntityMotionState.h [hunk 9]
```diff
@@ -100,6 +99,7 @@ class EntityMotionState : public ObjectMotionState {
     void saveKinematicState(btScalar timeStep) override;
 
 protected:
+    uint8_t computeFinalBidPriority() const;
```
Comment (AndrewMeadows) [inline libraries/physics/src/EntityMotionState.h]: This method was moved to `protected` status because it is only used for internal `EntityMotionState` logic.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16713/HighFidelity-Beta-PR14766-aff1244.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16713/HighFidelity-Beta-Interface-PR14766-aff1244.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16713/HighFidelity-Beta-PR14766-aff1244.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16713/HighFidelity-Beta-Interface-PR14766-aff1244.exe

**Comment by qliejahner** [comment]
Currently testing

**Comment by qliejahner** [comment]
Fail

Both bugs are not occurring; simulation ownership transfers with grabs, and non-dynamic entities can be grabbed and used to knock dynamic ones around.

However, while testing general grabbing mechanics, two critical problems appeared.

1) Non-dynamic entities cannot be moved with far grabs.
If another user near grabs the entity, it will stay in an average position between the two attempted grabs. However, if no user is near grabbing it, it cannot be moved.

hifi-snap-by-evanstaging-on-2019-01-24_12-57-31


2) Non-dynamic entities can be thrown, maintaining their velocity when released.

hifi-snap-by-evanstaging-on-2019-01-24_12-58-37


[hifi-log_2019-01-24_13.07.56.txt](https://github.com/highfidelity/hifi/files/2793433/hifi-log_2019-01-24_13.07.56.txt)
[hifi-log.txt](https://github.com/highfidelity/hifi/files/2793434/hifi-log.txt)

**Comment by AndrewMeadows** [comment]
**To test:**

**(1)**  Make a non-dynamic club and a dynamic floating ball.
**(2)** Enable this debug option: **Developers --> Physics --> Highlight Simulation Ownership**
**(3)** In HMD near-grab the bat --> notice that you will have simulation ownership.
**(4)** In HMD near-grab the bat an use it to hit the ball --> the ball should be activated in the physic simulation and should fly off as if it were hit by the club.  Also notice: you should have simulation ownership of the ball too.
**(5)** Note: if you swing the bat really really fast it is possible to tunnel it through the ball such that the ball does not move, or maybe moves but does not receive the full impact of the swing.  Unfortunately this is just how it is for fast-moving non-dynamic clubs.

**(6)** Note: step **(4)** is buggy for **far-grab** of non-dynamic club.  The ball will be activated in the physics simulation when the club hits it but the velocity it gets from the collision will tend to be wrong: it will be difficult to impart high velocity on the ball.  This is known and will not be fixed in this PR.  The problem here is the FarGrab action doesn't know how to operate on a kinematic object yet, so we fall back to just teleporting the club to its new location without setting its velocities to agree with it effective motion.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16747/HighFidelity-Beta-PR14766-c9ef439.apk

File: libraries/avatars/src/AvatarData.cpp [hunk 10]
```diff
@@ -2997,7 +2997,6 @@ void AvatarData::clearAvatarGrabData(const QUuid& grabID) {
     _avatarGrabsLock.withWriteLock([&] {
         if (_avatarGrabData.remove(grabID)) {
             _avatarGrabDataChanged = true;
-            _deletedAvatarGrabs.insert(grabID);
         }
     });
```
Comment (AndrewMeadows) [inline libraries/avatars/src/AvatarData.cpp]: This fix closes a resource leak on the **avatar-mixer**.  Whenever we would `clearAvatarGrabData()` a new `grabID` would be pushed on the `_deletedAvatarGrabs` set... and it would never be cleared out because that was only done in the `Avatar` class which lives on the **interface**.  I moved that container to `Avatar` (and changed its name and type) for Great Justice.

File: libraries/avatars/src/AvatarData.h [hunk 11]
```diff
@@ -54,7 +54,6 @@
 #include "AvatarTraits.h"
 #include "HeadData.h"
 #include "PathUtils.h"
-#include "Grab.h"
```
Comment (AndrewMeadows) [inline libraries/avatars/src/AvatarData.h]: TODO: Now that `AvatarData` really doesn't need to know about the `Grab` struct we need to move `Grab.h` out of **avatars** lib and into **interface**.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16747/HighFidelity-Beta-PR14766-c9ef439.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16747/HighFidelity-Beta-Interface-PR14766-c9ef439.dmg

**Comment by qliejahner** [comment]
Currently testing

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16799/HighFidelity-Beta-PR14766-f2b9f08.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16799/HighFidelity-Beta-PR14766-f2b9f08.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16799/HighFidelity-Beta-Interface-PR14766-f2b9f08.exe

**Comment by qliejahner** [comment]
Currently testing

**Comment by qliejahner** [comment]
PASS

Original and subsequent bugs are fixed! No new bad behaviors were seen
